File: UniCalib/unicalib/utils/transforms.py
"""
几何变换工具函数
"""
from __future__ import annotations
from typing import List

import numpy as np
from scipy.spatial.transform import Rotation


def rotation_matrix_to_euler(R: np.ndarray, degrees: bool = True) -> np.ndarray:
    """3×3 旋转矩阵 → ZYX 欧拉角。"""
    return Rotation.from_matrix(R).as_euler("zyx", degrees=degrees)


def euler_to_rotation_matrix(euler_zyx: np.ndarray, degrees: bool = True) -> np.ndarray:
    """ZYX 欧拉角 → 3×3 旋转矩阵。"""
    return Rotation.from_euler("zyx", euler_zyx, degrees=degrees).as_matrix()


def skew_symmetric(v: np.ndarray) -> np.ndarray:
    """3-向量 → 3×3 反对称矩阵。"""
    return np.array([
        [0.0, -v[2], v[1]],
        [v[2], 0.0, -v[0]],
        [-v[1], v[0], 0.0],
    ])


def rotation_distance_deg(R1: np.ndarray, R2: np.ndarray) -> float:
    """计算两个旋转矩阵之间的角度距离（°）。"""
    dR = R1.T @ R2
    trace = np.clip((np.trace(dR) - 1.0) / 2.0, -1.0, 1.0)
    return float(np.degrees(np.arccos(trace)))


def quaternion_mean(quats: np.ndarray) -> np.ndarray:
    """
    计算四元数集合的平均值（特征向量法）。
    quats: (N, 4) 格式 [x, y, z, w]
    """
    Q = np.array(quats)
    M = Q.T @ Q
    eigvals, eigvecs = np.linalg.eigh(M)
    return eigvecs[:, np.argmax(eigvals)]


def T_from_Rt(R: np.ndarray, t: np.ndarray) -> np.ndarray:
    """拼合 4×4 变换矩阵 T = [R | t; 0 | 1]。"""
    T = np.eye(4)
    T[:3, :3] = R
    T[:3, 3] = t.flatten()
    return T


def Rt_from_T(T: np.ndarray):
    """从 4×4 变换矩阵分解 R, t。"""
    return T[:3, :3], T[:3, 3]


def robust_rotation_average(rotations: List[np.ndarray],
                             n_ransac: int = 100,
                             threshold_deg: float = 5.0) -> np.ndarray:
    """
    RANSAC + 四元数平均 鲁棒旋转集成。
    """
    if not rotations:
        return np.eye(3)
    if len(rotations) == 1:
        return rotations[0]

    quats = np.array([
        Rotation.from_matrix(R).as_quat() for R in rotations
    ])

    best_inliers: List[int] = []
    rng = np.random.default_rng(42)

    for _ in range(n_ransac):
        idx = rng.integers(0, len(quats))
        ref = quats[idx]
        angles = np.array([
            float(np.degrees(2.0 * np.arccos(
                np.clip(abs(float(np.dot(ref, q))), 0.0, 1.0))))
            for q in quats
        ])
        inliers = [i for i, a in enumerate(angles) if a < threshold_deg]
        if len(inliers) > len(best_inliers):
            best_inliers = inliers

    if not best_inliers:
        best_inliers = list(range(len(quats)))

    mean_q = quaternion_mean(quats[best_inliers])
    return Rotation.from_quat(mean_q).as_matrix()


def interpolate_se3(T1: np.ndarray, T2: np.ndarray, alpha: float) -> np.ndarray:
    """在两个 SE3 变换之间线性插值（slerp 旋转 + 线性平移）。"""
    R1, t1 = Rt_from_T(T1)
    R2, t2 = Rt_from_T(T2)

    rot1 = Rotation.from_matrix(R1)
    rot2 = Rotation.from_matrix(R2)
    # SLERP
    q1, q2 = rot1.as_quat(), rot2.as_quat()
    dot = np.dot(q1, q2)
    if dot < 0:
        q2 = -q2
        dot = -dot
    dot = np.clip(dot, -1.0, 1.0)
    theta = np.arccos(dot)
    if abs(theta) < 1e-8:
        q_interp = q1
    else:
        q_interp = (np.sin((1 - alpha) * theta) * q1 +
                    np.sin(alpha * theta) * q2) / np.sin(theta)

    R_interp = Rotation.from_quat(q_interp).as_matrix()
    t_interp = (1 - alpha) * t1 + alpha * t2

    return T_from_Rt(R_interp, t_interp)

File: calib_unified/src/unicalib_legacy/transforms.cpp
/**
 * UniCalib C++ — 几何变换实现 (ZYX 欧拉角、四元数、SE3)
 */
#include "unicalib/transforms.hpp"
#include <cmath>
#include <random>
#include <algorithm>

namespace unicalib {

namespace {

const double kPi = 3.14159265358979323846;

// 旋转矩阵 → 四元数 [x,y,z,w]
Eigen::Vector4d rotation_to_quat(const Eigen::Matrix3d& R) {
  Eigen::Quaterniond q(R);
  return Eigen::Vector4d(q.x(), q.y(), q.z(), q.w());
}

Eigen::Matrix3d quat_to_rotation(const Eigen::Vector4d& q) {
  return Eigen::Quaterniond(q(3), q(0), q(1), q(2)).toRotationMatrix();
}

}  // namespace

Eigen::Vector3d rotation_matrix_to_euler(const Eigen::Matrix3d& R, bool degrees) {
  double sy = std::sqrt(R(0, 0) * R(0, 0) + R(1, 0) * R(1, 0));
  bool singular = sy < 1e-8;
  double x, y, z;
  if (!singular) {
    x = std::atan2(R(2, 1), R(2, 2));
    y = std::atan2(-R(2, 0), sy);
    z = std::atan2(R(1, 0), R(0, 0));
  } else {
    x = std::atan2(-R(1, 2), R(1, 1));
    y = std::atan2(-R(2, 0), sy);
    z = 0.0;
  }
  Eigen::Vector3d euler(x, y, z);
  if (degrees) euler *= (180.0 / kPi);
  return euler;
}

Eigen::Matrix3d euler_to_rotation_matrix(const Eigen::Vector3d& euler_zyx, bool degrees) {
  double x = euler_zyx(0), y = euler_zyx(1), z = euler_zyx(2);
  if (degrees) {
    x *= kPi / 180.0;
    y *= kPi / 180.0;
    z *= kPi / 180.0;
  }
  Eigen::AngleAxisd rx(x, Eigen::Vector3d::UnitX());
  Eigen::AngleAxisd ry(y, Eigen::Vector3d::UnitY());
  Eigen::AngleAxisd rz(z, Eigen::Vector3d::UnitZ());
  return (rz * ry * rx).toRotationMatrix();
}

Eigen::Matrix3d skew_symmetric(const Eigen::Vector3d& v) {
  Eigen::Matrix3d S;
  S << 0, -v(2), v(1), v(2), 0, -v(0), -v(1), v(0), 0;
  return S;
}

double rotation_distance_deg(const Eigen::Matrix3d& R1, const Eigen::Matrix3d& R2) {
  Eigen::Matrix3d dR = R1.transpose() * R2;
  double trace = (dR.trace() - 1.0) / 2.0;
  trace = std::max(-1.0, std::min(1.0, trace));
  return std::acos(trace) * (180.0 / kPi);
}

Eigen::Vector4d quaternion_mean(const std::vector<Eigen::Vector4d>& quats) {
  if (quats.empty()) return Eigen::Vector4d(0, 0, 0, 1);
  Eigen::Matrix4d M = Eigen::Matrix4d::Zero();
  for (const auto& q : quats) M += q * q.transpose();
  Eigen::SelfAdjointEigenSolver<Eigen::Matrix4d> es(M);
  int idx = 0;
  double max_val = es.eigenvalues()(0);
  for (int i = 1; i < 4; ++i) {
    if (es.eigenvalues()(i) > max_val) {
      max_val = es.eigenvalues()(i);
      idx = i;
    }
  }
  return es.eigenvectors().col(idx);
}

Eigen::Matrix4d T_from_Rt(const Eigen::Matrix3d& R, const Eigen::Vector3d& t) {
  Eigen::Matrix4d T = Eigen::Matrix4d::Identity();
  T.block<3, 3>(0, 0) = R;
  T.block<3, 1>(0, 3) = t;
  return T;
}

void Rt_from_T(const Eigen::Matrix4d& T, Eigen::Matrix3d& R, Eigen::Vector3d& t) {
  R = T.block<3, 3>(0, 0);
  t = T.block<3, 1>(0, 3);
}

Eigen::Matrix3d robust_rotation_average(
  const std::vector<Eigen::Matrix3d>& rotations,
  int n_ransac,
  double threshold_deg) {
  if (rotations.empty()) return Eigen::Matrix3d::Identity();
  if (rotations.size() == 1) return rotations[0];

  std::vector<Eigen::Vector4d> quats;
  quats.reserve(rotations.size());
  for (const auto& R : rotations) quats.push_back(rotation_to_quat(R));

  std::vector<int> best_inliers;
  std::mt19937 rng(42);
  std::uniform_int_distribution<int> dist(0, static_cast<int>(quats.size()) - 1);

  for (int iter = 0; iter < n_ransac; ++iter) {
    int idx = dist(rng);
    const Eigen::Vector4d& ref = quats[idx];
    std::vector<int> inliers;
    for (size_t i = 0; i < quats.size(); ++i) {
      double dot = std::abs(ref.dot(quats[i]));
      dot = std::max(0.0, std::min(1.0, dot));
      double angle_deg = std::acos(dot) * 2.0 * (180.0 / kPi);
      if (angle_deg < threshold_deg) inliers.push_back(static_cast<int>(i));
    }
    if (inliers.size() > best_inliers.size()) best_inliers = inliers;
  }

  if (best_inliers.empty())
    for (size_t i = 0; i < quats.size(); ++i) best_inliers.push_back(static_cast<int>(i));

  std::vector<Eigen::Vector4d> inlier_quats;
  for (int i : best_inliers) inlier_quats.push_back(quats[i]);
  Eigen::Vector4d mean_q = quaternion_mean(inlier_quats);
  return quat_to_rotation(mean_q);
}

Eigen::Matrix4d interpolate_se3(
  const Eigen::Matrix4d& T1,
  const Eigen::Matrix4d& T2,
  double alpha) {
  Eigen::Matrix3d R1, R2;
  Eigen::Vector3d t1, t2;
  Rt_from_T(T1, R1, t1);
  Rt_from_T(T2, R2, t2);

  Eigen::Vector4d q1 = rotation_to_quat(R1);
  Eigen::Vector4d q2 = rotation_to_quat(R2);
  double dot = q1.dot(q2);
  if (dot < 0) { q2 = -q2; dot = -dot; }
  dot = std::max(-1.0, std::min(1.0, dot));
  double theta = std::acos(dot);

  Eigen::Vector4d q_interp;
  if (std::abs(theta) < 1e-8) {
    q_interp = q1;
  } else {
    q_interp = (std::sin((1 - alpha) * theta) * q1 + std::sin(alpha * theta) * q2) / std::sin(theta);
  }
  Eigen::Matrix3d R_interp = quat_to_rotation(q_interp);
  Eigen::Vector3d t_interp = (1 - alpha) * t1 + alpha * t2;
  return T_from_Rt(R_interp, t_interp);
}

}  // namespace unicalib
